Blog contribution: article about on-demand shared graphql subscriptions

gthau opened the pull request:
Hi The Guild!

I'd like to propose this blog article. It's about solving for an issue related to subscriptions whose values are expensive to compute. The solution describes how to:
- start computing updates on-demand
- share the subscription's values between several subscribers
- stop (with or without delay) the computation of those updates

All without relying on too much explicit state (no counting of subscribers), relying on the power of RXJS.

Let me know what you think :)

dotansimha commented:
Thanks @gthau , sorry for the delay in reviewing this. Will take a look soon :)

gthau commented:
> Thanks @gthau , sorry for the delay in reviewing this. Will take a look soon :)

No problem. 

I will to add a paragraph to the article about the ability to run logic on subscription close, which is something that's not possible out of the box with graphql (since we have only subscribe and resolve functions).

n1ru4l reviewed (approved):
Really good high quality article 😍

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+---
+title: On-demand shared subscriptions using RXJS
```
```suggestion
title: On-demand shared subscriptions using RxJS
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+
+The solution to the simple example is implemented in this [repository](https://github.com/gthau/graphql-shared-subscriptions-with-rxjs).
+
+### Using RXJS multicasting capabilities
```
```suggestion
### Using RxJS multicasting capabilities
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
````diff
@@ -0,0 +1,335 @@
+queryPromise.then(console.log);
+```
+
+However RXJS has multicasting capabilities to turn cold observables into hot observables. Using RXJS's multicasting operators, we can create observable pipelines which are shared: their side-effect is triggered when the first subscriber subscribes to it and when more than one subscriber subscribe, the pipeline is executed only once and the emitted value is passed to all subscribers.
````
```suggestion
However, RxJS has multicasting capabilities to turn cold observables into hot observables. Using RxJS's multicasting operators, we can create observable pipelines which are shared: their side-effect is triggered when the first subscriber subscribes to it and when more than one subscriber subscribes, the pipeline is executed only once and the emitted value is passed to all subscribers.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+- we can have one user subscribing to the view, this user refreshes his browser's page, effectively ending the subscription, triggering a cache cleanup, but then the view is resubscribed almost immediately, triggering a new fresh fetch
+- we can have one user stopping the subscription (e.g. navigating away from this view to another part of our app), but another user subscribing soon (a few milliseconds, seconds or minutes) after
+
+To optimize for those use cases, we can delay the unsubscription of the shared observable when no more subscriber are subscribing to it, using the `resetOnRefCountZero` config of the `share` operator. This is available in RXJS v7.
```
```suggestion
To optimize for those use cases, we can delay the unsubscription of the shared observable when no more subscribers are subscribing to it, using the `resetOnRefCountZero` config of the `share` operator. This is available in RxJS v7.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+In order to execute logic, we have 2 places: in `subscribe` when the subscription starts and in `resolve` on each emitted value. What if we want to execute logic when the subscription ends (whether it completed or was stopped by the client)?
+
+The answer is: it is not contemplated in the spec. But it doesn't mean we can't do it.
+In fact it is even very simple! If we use the solution described above, we already have all the pieces in place. We only need to add one line: in the `subscribe` function where we set up the AsyncIterable (by transforming the Observable into an AsyncIterable), we only have to use RXJS's `finalize` operator.
```
```suggestion
In fact it is even very simple! If we use the solution described above, we already have all the pieces in place. We only need to add one line: in the `subscribe` function where we set up the AsyncIterable (by transforming the Observable into an AsyncIterable), we only have to use RxJS's `finalize` operator.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+
+### Buffer events
+
+First we'll buffer the events, dedupe them and share the observable because those events will be checked against every playlist subscribed to, so we can share the buffering and deduplication:
```
```suggestion
First, we'll buffer the events, dedupe them and share the observable because those events will be checked against every playlist subscribed to, so we can share the buffering and deduplication:
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+- others are related to the parent multimedia objects (object deleted, object's metadata changed, object has new children objects)
+- others are related to the children objects of the top level objects (object deleted, object's metadata changed)
+
+This playlist's data is expensive to fetch, can be large and is cached only when users subscribe to it: because it is expensive to compute updates, we want to ensure we don't process events for all playlists in the system, only the ones some user is currently interested in.
```
```suggestion
This playlist's data is expensive to fetch, can be large, and is cached only when users subscribe to it: because it is expensive to compute updates, we want to ensure we don't process events for all playlists in the system, only the ones some user is currently interested in.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+Let's consider a view such as a complex "playlist": this playlist is a list of items organized in blocks, the items themselves are a complex aggregate of several multimedia (video, audio, graphics) objects, those multimedia objects themselves might be aggregate (bundles) or have of some children objects (e.g. sub-clips, graphics on a video, subtitles). In effect this complex playlist is organized ...
+
+The system emits different events:
+- some are related directly to the playlist itself (e.g. new item in playlist, item removed from playlist, item changed position inside playlist)
```
```suggestion
- some are related directly to the playlist itself (e.g. new item in the playlist, item removed from the playlist, item changed position inside playlist)
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+
+## Apply to a real-life example
+
+Let's consider a view such as a complex "playlist": this playlist is a list of items organized in blocks, the items themselves are a complex aggregate of several multimedia (video, audio, graphics) objects, those multimedia objects themselves might be aggregate (bundles) or have of some children objects (e.g. sub-clips, graphics on a video, subtitles). In effect this complex playlist is organized ...
```
```suggestion
Let's consider a view such as a complex "playlist": this playlist is a list of items organized in blocks, the items themselves are a complex aggregate of several multimedia (video, audio, graphics) objects, those multimedia objects themselves might be aggregate (bundles) or have of some children objects (e.g. sub-clips, graphics on a video, subtitles). In effect, this complex playlist is organized in the form of a tree of objects.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
````diff
@@ -0,0 +1,335 @@
+}
+```
+
+In the previous examples, we used the `finalize` operator on the shared observable pipeline in order to clean up after all subscriptions to a specific entity had ended. But we can also use it per-subscription by piping it onto the shared observable: this way it is only run for when this specific subscription ends.
````
```suggestion
In the previous examples, we used the `finalize` operator on the shared observable pipeline in order to clean up after all subscriptions to a specific entity had ended. But we can also use it per subscription by piping it onto the shared observable: this way it is only run for when this specific subscription ends.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+### Run server-side logic when a user's subscription is closed
+
+In GraphQL we implement subscriptions by writing two functions:
+- a `subscribe` function which is run one time only and must returns an AsyncIterable
```
```suggestion
- a `subscribe` function which is run one time only and must return an AsyncIterable
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+In the example above, after all subscribers have unsubscribed (`refcount` is 0) the connector (the subject that subscribes to the source observable) now waits for 10 seconds (10000 ms) before it resets the shared observable. During this period of time, the observable pipeline continues to execute. If a new subscriber subscribes before the `timer` has emitted, the `refcount` is again greater than  ...
+
+<Callout type="warning">
+  We use `timer` in this example but we can use any observable (e.g. receiving a specific event from our message queue) that emits at some point. Be careful though to make sure that the observable you use will actually emit otherwise you'll have a memory leak. To remedy this issue, you might want to add a safety, e.g. using [`raceWith`](https://rxjs.dev/api/operators/raceWith) with `timer` to mak ...
```
```suggestion
  We use `timer` in this example but we can use any observable (e.g. receiving a specific event from our message queue) that emits at some point. Be careful though to make sure that the observable you use will actually emit otherwise you'll have a memory leak. To remedy this issue, you might want to add safety, e.g. using [`raceWith`](https://rxjs.dev/api/operators/raceWith) with `timer` to make sure you don't wait indefinitely:
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+### Delay the shared observable cleanup
+
+Because the view might be expensive to fetch and update, we only do so when users are subscribing to it. In the previous example, as soon as all users have unsubscribed, the shared observable is unsubscribed and deleted from the cache. This might not be the desired behavior, e.g.:
+- we can have one user subscribing to the view, this user refreshes his browser's page, effectively ending the subscription, triggering a cache cleanup, but then the view is resubscribed almost immediately, triggering a new fresh fetch
```
```suggestion
- we can have one user subscribing to the view, this user refreshes his browser's page, effectively ending the subscription, and triggering a cache cleanup, but then the view is resubscribed almost immediately, triggering a new fresh fetch
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+  The graphql-js engine expects the `subscribe` handler to return an AsyncIterable, so we use the [`rxjs-for-await` package from Ben Lesh](https://github.com/benlesh/rxjs-for-await) which contains different functions from transforming an Observable into an AsyncIterable (there are different functions to deal with backpressure and lossiness/losslessness).
+</Callout>
+
+So far, so good. But we have a memory leak: after all subscribers have unsubscribed, we keep the publishers in the cache. We can clean those up using the [`finalize`](https://rxjs.dev/api/operators/finalize) operator: this operator allows to run a callback on observable termination (either the observable is complete or has no more subscribers).
```
```suggestion
So far, so good. But we have a memory leak: after all subscribers have unsubscribed, we keep the publishers in the cache. We can clean those up using the [`finalize`](https://rxjs.dev/api/operators/finalize) operator: this operator allows running a callback on observable termination (either the observable is complete or has no more subscribers).
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
````diff
@@ -0,0 +1,335 @@
+
+```
+
+Now from the our GraphQL's subscription resolver, we can use it:
````
```suggestion
Now from our GraphQL's subscription resolver, we can use it like the following:
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+The solution to the simple example is implemented in this [repository](https://github.com/gthau/graphql-shared-subscriptions-with-rxjs).
+
+### Using RXJS multicasting capabilities
+By default observables are lazy and cold, meaning their side-effect are run only when a subscriber subscribes to the observable and for every subscriber the observable pipeline is executed.
```
```suggestion
By default observables are lazy and cold, meaning their side-effect are run only when a subscriber subscribes to the observable. Furthermore, for every subscriber, the observable pipeline is executed.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+
+However this ideal scenario is not always possible. We might have "complex" entities we want to subscribe to, e.g. views that are aggregates of data over several data sources. The view might be expensive to fetch and subscribing to the view's content updates might involve subscribing to multiple events and perform expensive processing and/or additional fetching in order to compute the updates and ...
+
+In this case, if multiple subscribers subscribe to the same view, the events' processing, additional fetching and computing would have to be done for each subscriber, which will lead to performance issues as the number of subscribers increases.
```
```suggestion
In this case, if multiple subscribers subscribe to the same view, the events' processing, additional fetching, and computing would have to be done for each subscriber, which will lead to performance issues as the number of subscribers increases.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+
+Ideally we want subscriptions to be stateless and independent for every subscriber: when a subscriber subscribes to an entity, this subscription is independent from any other subscription and can be handled by any instance of our GraphQL server. This is the case when the entity subscribed to is "simple" and processing updates is trivial: for example, we receive events from a pubsub provider / mes ...
+
+However this ideal scenario is not always possible. We might have "complex" entities we want to subscribe to, e.g. views that are aggregates of data over several data sources. The view might be expensive to fetch and subscribing to the view's content updates might involve subscribing to multiple events and perform expensive processing and/or additional fetching in order to compute the updates and ...
```
```suggestion
However, this ideal scenario is not always possible. We might have "complex" entities we want to subscribe to, e.g. views that are aggregates of data over several data sources. The view might be expensive to fetch and subscribing to the view's content updates might involve subscribing to multiple events and performing expensive processing and/or additional fetching in order to compute the updates and send them to the subscriber.
```

n1ru4l commented on website/pages/blog/graphql-shared-stateful-subscription-with-rxjs.mdx:
```diff
@@ -0,0 +1,335 @@
+
+GraphQL has native support for subscriptions as schema operations. GraphQL subscriptions allow to receive continuous updates, usually implemented using a persistent WebSocket connection or Server-Sent Events.
+
+Ideally we want subscriptions to be stateless and independent for every subscriber: when a subscriber subscribes to an entity, this subscription is independent from any other subscription and can be handled by any instance of our GraphQL server. This is the case when the entity subscribed to is "simple" and processing updates is trivial: for example, we receive events from a pubsub provider / mes ...
```
```suggestion
Ideally, we want subscriptions to be stateless and independent for every subscriber: when a subscriber subscribes to an entity, this subscription is independent of any other subscription and can be handled by any instance of our GraphQL server. This is the case when the entity subscribed to is "simple" and processing updates is trivial: for example, we receive events from a pubsub provider/message queue (e.g. Redis, Nats.io, etc.) in the form ```{ type: update, entityId: xyz }```. In this case, our subscription handler can simply subscribe to these events, filter on the requested `entityId`, and forward those events (maybe a bit processed) to the subscriber.
```

n1ru4l commented:
@gthau Would you be fine with simplifying the blog post name a bit too `Shared GraphQL Subscription Sources with RxJS`?

n1ru4l commented:
<img width="844" alt="shared rxjs subscription" src="https://user-images.githubusercontent.com/14338007/211826594-625f70f4-25f4-49c6-b89b-79309532b437.png">

<img width="328" alt=" shared rxjs subscription" src="https://user-images.githubusercontent.com/14338007/211826788-c063db5a-0413-4de4-afb1-5b38925c4192.png">

gthau commented:
> @gthau Would you be fine with simplifying the blog post name a bit to `Shared GraphQL Subscription Sources with RxJS`?

This shortened post name misses one of the two main point of the post which is how to perform (start and stop) heavy processing only when there are subscribers. That's what I wanted to convey using the word "On-Demand", although I'm not sure it gets the idea across.

n1ru4l commented:
@gthau this will look really packed though

<img width="542" alt="image" src="https://user-images.githubusercontent.com/14338007/212036667-25f1327e-ac76-45d1-b6be-2ec5fac2310b.png">

gthau commented:
> @gthau this will look really packed though

Yes... Let's go with the shortened post name then.

I'll add a sentence in the introduction to make it clear from the get-go what it is we want to achieve.

enisdenjo reviewed (commented):
This looks nice, great job!

Only one thing I'd recommend: include the import statements. A question readers might ask is where do the `interval`, `share`, `eachValueFrom`, and each of the RxJS operators come from?

gthau commented:
> This looks nice, great job!
> 
> Only one thing I'd recommend: include the import statements. A question readers might ask is where do the `interval`, `share`, `eachValueFrom`, and each of the RxJS operators come from?

Good suggestion, will do it tomorrow.

Do you think a bit more explanation is required about observables, lazy vs eager, rxjs in general? RxJS is well known to Angular devs, but less so to React and GraphQL backend devs. But as the article demonstrates, it has more use cases than just being an Angular implementation detail and can solve problems in elegant ways.

enisdenjo commented:
> Do you think a bit more explanation is required about observables, lazy vs eager, rxjs in general? RxJS is well known to Angular devs, but less so to React and GraphQL backend devs.

I've used RxJS and the Observable pattern before, outside Angular. The way I found it useful is almost exact to this post, how you can split and share streams. But, I don't think we should overdo the RxJS explanations here - those curious about the intricacies will do the googling. Maybe just a few words about the Observable pattern and that'd be it. What do you say @gthau?

gthau commented:
> > Do you think a bit more explanation is required about observables, lazy vs eager, rxjs in general? RxJS is well known to Angular devs, but less so to React and GraphQL backend devs.
> 
> I've used RxJS and the Observable pattern before, outside Angular. The way I found it useful is almost exact to this post, how you can split and share streams. But, I don't think we should overdo the RxJS explanations here - those curious about the intricacies will do the googling. Maybe just a few words about the Observable pattern and that'd be it. What do you say @gthau?

Agreed, we'll do tomorrow.

gthau commented:
> > Do you think a bit more explanation is required about observables, lazy vs eager, rxjs in general? RxJS is well known to Angular devs, but less so to React and GraphQL backend devs.
> 
> I've used RxJS and the Observable pattern before, outside Angular. The way I found it useful is almost exact to this post, how you can split and share streams. But, I don't think we should overdo the RxJS explanations here - those curious about the intricacies will do the googling. Maybe just a few words about the Observable pattern and that'd be it. What do you say @gthau?

I added the import statements in the code snippets and added a word of RxJS.

enisdenjo commented:
Love it, awesome!

gthau commented:
Sorry I added one more paragraph about sharing (base) sources across different graphql subscription operations (not only a single graphql subscription with same argument). I forgot this use case (that's the final thing I'll add, I promise :))

gthau commented:
Is there left anything to correct?

How can I see a rendered version of the article locally on my laptop?

enisdenjo commented:
> How can I see a rendered version of the article locally on my laptop?

1. `pnpm i`
2. `cd website`
3. `pnpm dev`

n1ru4l commented:
<img width="844" alt="shared rxjs subscription" src="https://user-images.githubusercontent.com/14338007/214263977-e6a261e6-7f64-42f7-bc8b-1c4a0401229e.png">
<img width="328" alt=" shared rxjs subscription" src="https://user-images.githubusercontent.com/14338007/214263993-c837046a-5654-4902-b4a2-3aba780a2f9a.png">
